### ISSUE TITLE
GovCloud CLI InvalidToken Error

### ISSUE THREAD

**Issue by CGick** [aws/aws-cli#5564]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [x] I've gone though the [User Guide](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-welcome.html) and the [API reference](https://docs.aws.amazon.com/cli/latest/reference/)
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-cli/issues) and didn't find any solution
  
**Describe the bug**
When attempting to run `aws s3 ls` on EC2 instance in `us-gov-west-1`, I recieved the following error 
```
aws s3 ls 

An error occurred (InvalidToken) when calling the ListBuckets operation: The provided token is malformed or otherwise invalid.
```

After doing some troubleshooting I discovered that the `--region us-gov-west-1` parameter fixed this error message.

**SDK version number**
```
aws --version
aws-cli/1.16.300 Python/2.7.16 Linux/4.14.171-136.231.amzn2.x86_64 botocore/1.13.36
```
**Platform/OS/Hardware/Device**
EC2 Amazon Linux2 instance

**To Reproduce (observed behavior)**
On an EC2 instance in the AWS GovCloud run the command `aws s3 ls`

**Expected behavior**
This error should be handled more clearly with a message that suggests using the region parameter.

**Comment by kdaily** [comment]
Hi @CGick,

It looks like this is an expected behavior. As noted in the GovCloud documentation, you must specify a region:

https://docs.aws.amazon.com/govcloud-us/latest/UserGuide/cli-and-api-access.html

This error is coming from the server-side API. I'll pass along this report to the S3 team about the error messaging. Other endpoint errors throw a `IllegalLocationConstraintException`, which per the [error code description](https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html) says:

> Indicates that you are attempting to access a bucket from a different region than where the bucket exists. To avoid this error, use the `--region` option. For example: `aws s3 cp awsexample.txt s3://testbucket/ --region ap-east-1`.

**Comment by ElijahLynn** [comment]
Yeah, this is a bad error message and needs to be fixed. Can we re-open this issue until it is fixed on the S3 side?

Crosslinking this issue where at least 14 others have run into this issue with the misleading error message and it actually needing a region. https://stackoverflow.com/q/54837248/292408

**Comment by ElijahLynn** [comment]
I think the `aws s3` code could catch that response and append some text to the exception saying this could also be caused by a missing region, at the least. So I do think this should be re-opened and handled better by this tool. Thanks

**Comment by ElijahLynn** [comment]
Some starting points on the code:

https://github.com/aws/aws-cli/blob/develop/awscli/customizations/s3errormsg.py
https://github.com/aws/aws-cli/blob/develop/awscli/customizations/s3/

**Comment by olivereri** [comment]
> Yeah, this is a bad error message and needs to be fixed. Can we re-open this issue until it is fixed on the S3 side?
> 
> Crosslinking this issue where at least 14 others have run into this issue with the misleading error message and it actually needing a region. https://stackoverflow.com/q/54837248/292408

I agree, may we please have this issue re-opened?  `InvalidToken`,  `The provided token is malformed or otherwise invalid.` or other messages of its kind are too cryptic making it hard to find a solution.

**Comment by aarowman** [comment]
Just ran into this today - please reopen. The issue is not the token, but the region. The error message is very misleading and causes lost time troubleshooting for the wrong reason

**Comment by CerebralMastication** [comment]
I ran into this exact issue today. This error message is garbage and could be trivially improved.

**Comment by mgipson** [comment]
Hi there, I think this may still be an issue.

**Describe the bug**
When attempting to run aws s3 ls on bucket in us-east-1 using a saved profile with short-term access keys, I received the following error: 
```
aws s3 ls --profile test
An error occurred (InvalidToken) when calling the ListBuckets operation: The provided token is malformed or otherwise invalid.
```
But, `aws s3 --region us-east-1 ls --profile test` successfully lists the S3 buckets.

The `~/.aws/config` file looks like:
```
[profile test]
region = us-east-1
output = json
```
Tried changing `[profile test]` to `[test]` and `[default]`, but did not change behavior.

The `~/.aws/credentials` file looks like (containing short-term access keys):
```
[test]
aws_access_key_id=***
aws_secret_access_key=***
aws_session_token=***
```

**SDK version number**
aws --version
aws-cli/2.3.7 Python/3.8.8 Darwin/22.6.0 exe/x86_64 prompt/off

**Platform/OS/Hardware/Device**
Local machine (Mac, Ventura 13.6)

**To Reproduce (observed behavior)**
On local machine (Mac), run the command `aws s3 ls --profile myprofile` with a saved profile of short-term access keys in `~/.aws/credentials`.

**Expected behavior**
This error should be handled more clearly with a message that suggests using the region parameter, or use the region parameter indicated in `~/.aws/config` per the documentation.

**Comment by tim-finnigan** [comment]
Hi all, thanks for your patience here. I'm reopening this issue and transferring it to our cross-SDK repository for further tracking. [This earlier comment](https://github.com/aws/aws-cli/issues/5564#issuecomment-701610980) summarizes the issue:

> It looks like this is an expected behavior. As noted in the GovCloud documentation, you must specify a region:
> 
> https://docs.aws.amazon.com/govcloud-us/latest/UserGuide/cli-and-api-access.html
> 
> This error is coming from the server-side API. I'll pass along this report to the S3 team about the error messaging. Other endpoint errors throw a `IllegalLocationConstraintException`, which per the [error code description](https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html) says:
> 
>  >  Indicates that you are attempting to access a bucket from a different region than where the bucket exists. To avoid this error, use the `--region` option. For example: `aws s3 cp awsexample.txt s3://testbucket/ --region ap-east-1`.

To highlight that this issue is with the S3 API and not directly the `s3` customization commands, the `s3api` calls will fail with the same error, for example: `aws s3api <command> --region us-gov-west-1`

I'll reach out to the S3 team and try to get some traction on this.

---

@mgipson what you reported looks like a different issue. Version 2.3.7 is quite old and I don't see that issue on v2.13.38. Please try [updating](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) to a newer version of the CLI and let us know if you see the same issue after that.
